File: lengthOfLIS.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        '''
        T = O(NlogN) , N -> length of the nums
        S = O(N)
        
        '''
        if not nums: return 0 
        n = len(nums)
        dp = [0]*n
        dp[0] = nums[0]
        len_ = 1 
        for i in range(1, n):
            if nums[i] > dp[len_ -1]:
                dp[len_] = nums[i]
                len_ += 1 
            else:
                idx = self.binary_search(dp, nums[i] , 0 , len_ - 1 )
                dp[idx] = nums[i] 
        return len_
	
    def binary_search(self, nums,target,low,high):
        while low <= high:
            mid = low + (high - low) // 2
            if nums[mid] == target:
                return mid
            elif target < nums[mid]:
                high = mid - 1
            else:
                low = mid + 1
        return low

    def subOptimallengthOfLIS(self, nums: List[int]) -> int:
        '''
        T = O(N^2) , N -> size of the input array 
        S = O(N)
        '''
        dp = [1]*len(nums)
        max_ = 1
        for i in range( 1 , len(nums) ):
            for  j in range(0 , i):
                if nums[j] < nums[i]:
                    dp[i] = max( dp[i] , dp[j] + 1 )
            max_ = max(max_ , dp[i])
        return max_
            
            

File: LengthOfLIS.java
// Time Complexity : The time complexity is O(nlogn) where n is the length of array
// Space Complexity : The space complexity is O(n) where n is the length of array
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {

    public int lengthOfLIS(int[] nums) {

        int n = nums.length;

        int[] dp = new int[n];
        dp[0] = nums[0];
        int index = 1;

        for(int i=1;i<n;i++){
            //if the current number forms an increasing sequence
            if(dp[index-1] < nums[i]){
                dp[index] = nums[i];
                index++;
            }
            // if the current number doesnt form an increasing sequence, do binary search to replace the closest next element
            else{
                int idx = binarySearch(dp,0,index-1,nums[i]);
                dp[idx] = nums[i];
            }
        }
        return index;
    }

    public int binarySearch(int[] nums,int start,int end,int target){

        while(start <= end){
            int mid = start + (end-start)/2;
            if(nums[mid] == target){
                return mid;
            }
            else if(nums[mid] > target){
                end = mid-1;
            }
            else{
                start = mid+1;
            }
        }
        return start;
    }
}